src/governance-validators.cpp

```diff
@@ -0,0 +1,349 @@
+        strErrorMessages += "name contains invalid characters;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateStartEndEpoch()
+{
+    int64_t nStartEpoch = 0;
+    int64_t nEndEpoch = 0;
+
+    if(!GetDataValue("start_epoch", nStartEpoch)) {
+        strErrorMessages += "start_epoch field not found;";
+        return false;
+    }
+
+    if(!GetDataValue("end_epoch", nEndEpoch)) {
+        strErrorMessages += "end_epoch field not found;";
+        return false;
+    }
+
+    if(nEndEpoch <= nStartEpoch) {
+        strErrorMessages += "end_epoch <= start_epoch field not found;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAmount()
+{
+    double dValue = 0.0;
+
+    if(!GetDataValue("payment_amount", dValue)) {
+        strErrorMessages += "payment_amount field not found;";
+        return false;
+    }
+
+    if(dValue <= 0.0) {
+        strErrorMessages += "payment_amount invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidatePaymentAddress()
+{
+    std::string strPaymentAddress;
+
+    if(!GetDataValue("payment_address", strPaymentAddress)) {
+        strErrorMessages += "payment_address field not found;";
+        return false;
+    }
+
+    static const std::string base58chars = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+    size_t nLength = strPaymentAddress.size();
+
+    if((nLength < 26) || (nLength > 35)) {
+        strErrorMessages += "incorrect payment_address length;";
+        return false;
+    }
+
+    if(strPaymentAddress.find_first_not_of(base58chars) != std::string::npos) {
+        strErrorMessages += "payment_address contains invalid characters;";
+        return false;
+    }
+
+    CBitcoinAddress address(strPaymentAddress);
+    if(!address.IsValid()) {
+        strErrorMessages += "payment_address is invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+bool CProposalValidator::ValidateURL()
+{
+    std::string strURL;
+    if(!GetDataValue("url", strURL)) {
+        strErrorMessages += "url field not found;";
+        return false;
+    }
+
+    std::string strURLStripped = StripWhitespace(strURL);
+
+    if(int(strURLStripped.size()) < 4) {
+        strErrorMessages += "url too short;";
+        return false;
+    }
+
+    if(!CheckURL(strURL)) {
+        strErrorMessages += "url invalid;";
+        return false;
+    }
+
+    return true;
+}
+
+void CProposalValidator::ParseJSONData()
+{
+    fJSONValid = false;
+
+    if(strData.empty()) {
+        return;
+    }
+
+    try {
+        UniValue obj(UniValue::VOBJ);
+        obj.read(strData);
+        std::vector<UniValue> arr1 = obj.getValues();
+        std::vector<UniValue> arr2 = arr1.at(0).getValues();
+        objJSON = arr2.at(1);
+        fJSONValid = true;
+    }
+    catch(std::exception& e) {
+        strErrorMessages += std::string(e.what()) + std::string(";");
```

Comment:
Compare generated code to estimate the cost in terms of code size.
The links below use Compiler Explorer https://godbolt.org,
(x86-64 gcc 6.3 with `-Os` option to optimise for size.)

For the baseline string concatenation with redundant string conversions:

       strErrorMessages += std::string(e.what()) + std::string(";");

[Baseline](https://godbolt.org/g/oyLBO2) >130 lines, 13 library calls

For the suggested refactor:

        strErrorMessages += e.what();
        strErrorMessages += ';';

[Suggestion](https://godbolt.org/g/6dtcR6) <10 lines 2 library calls

The cost of allocation is harder to quantify, Many small allocations add up.